### PR TITLE
SAMZA-1477: Fix issues found by BEAM tests

### PR DESCRIPTION
**Pull request by xinyuiscool**
A bunch of issues were found by BEAM tests, which includes:

1) WatermarkFunction needs to be able to return output after processWatermark()
2) control message doesn't implement the equals() and hashcode()
3) Some kafka system related code is not scala 2.10 compatible for tests.

### CODE REVIEW

File: samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.scala [hunk 1]
```diff
@@ -193,8 +193,8 @@ private[kafka] class KafkaSystemConsumer(
 
         // addTopicPartition one at a time, leaving the to-be-done list intact in case of exceptions.
         // This avoids trying to re-add the same topic partition repeatedly
-        def refresh(tp: List[TopicAndPartition]) = {
-          val head :: rest = tpToRefresh
+        def refresh() = {
```
Comment (prateekm) [inline samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.scala]: Why is this change required?

Comment (xinyuiscool) [reply samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.scala]: It doesn't compile with scala 2.10 in unit tests. Specifically, the scalaTest version we are using ("3.0.1") has some compatibility issues with 2.10.

File: samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.scala [hunk 2]
```diff
@@ -217,11 +217,11 @@ private[kafka] class KafkaSystemConsumer(
               case _ => debug("Ignoring refresh for %s because we already added it from another thread." format head)
             }
           }
-          rest
+          tpToRefresh.drop(1)
```
Comment (prateekm) [inline samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.scala]: Minor: Slightly cleaner to do tpToRefresh.tail. Is that equivalent here?

Comment (xinyuiscool) [reply samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.scala]: sure, that works too. Let me change to it.
